ledger/participant-state/kvutils/src/main/scala/kvutils/com/daml/ledger/participant-state/kvutils/InMemoryKVParticipantState.scala

```diff
@@ -0,0 +1,143 @@
+package com.daml.ledger.participant.state.kvutils
+
+import java.util.concurrent.atomic.AtomicReference
+
+import akka.NotUsed
+import akka.stream.scaladsl.Source
+import com.daml.ledger.participant.state.v1._
+import com.digitalasset.daml.lf.data.Ref.SimpleString
+import com.digitalasset.daml.lf.data.Time.Timestamp
+import com.digitalasset.daml.lf.engine.Engine
+import com.digitalasset.daml_lf.DamlLf.Archive
+import com.digitalasset.platform.akkastreams.Dispatcher
+import com.google.protobuf.ByteString
+
+import scala.collection.immutable.TreeMap
+import scala.concurrent.Future
+
+/* Implementation of the participant-state using the key-value utilities and an in-memory map. */
+class InMemoryKVParticipantState extends ReadService with WriteService {
+  val ledgerId = SimpleString.assertFromString("InMemoryKVParticipantState")
+
+  val engine = Engine()
+
+  val stateRef = new AtomicReference(
+    State(0, Timestamp.Epoch, TreeMap.empty, TreeMap.empty, Map.empty))
+  case class State(
+      nextIdx: Long,
+      recordTime: Timestamp,
+      // FIXME(JM): Add command de-duplication
+      entries: TreeMap[Long, (Timestamp, KeyValueUtils.KVEntryBlob)], // Tree map, so we can have ordered iteration.
+      rejections: TreeMap[Long, Update.CommandRejected],
```

Comment:
Why are the rejections not stored as `KVEntry` blobs as well? These rejections model rejections created during sequencing, which can only be communicated on-ledger.

Reply:
you're right, the rejections during sequencing need to go into the ledger.